### DISCUSSION TITLE
Problem with inputgen.py and setting up async mg-para calculations

### DISCUSSION THREAD

**Discussion by srdjan125** [Electrostatics/apbs#150] [Q&A]
Hello,

I would like to ask for help because I have several issues which are all connected to main problem:

I am trying to make APBS calculations on as large protein system, however after running psize.py program, I get recommendation to use parallel calculations with 300 processors.
 When I try to run mg-auto calculations with recommended cg/fg dimensions and grid points, calculations gets killed on my Ubuntu computer, while I can run it on cluster (but it is too slow considering the number of calculations I intend to perform).

I do not have access to resources to run parallel calculations with 300 processors, but I saw that it is possible to run several sequential calculations with mg-para and async. I have seen that running script inputgen.py can generate recommended input files, however I am having problem to understand hot to run that program.

In case of psize.py it is very easy: python2 psize.py File.pqr
however I cannot understand how to run inputgen.py script. Could someone help me with that issue?  
How the command should look like?    I have File.pqr and I want to run async calculations. 
I am trying like this but I am getting errors:

python inputgen.py File.pqr size.runPsize(Final.pqr) para 1

I would appreciate if someone could help me with this issue. It would be the best if I could see tutorial or how these async input files look like in order to understand how to adjust it for my case.

I also dont understand which apbs version I need to have in order to run async calculations? I am having problem to compile apbs with mpi and openmp in order to use mg-para but as I understood mg-para and async can be used without that "upgrade"?

Thank you in advance.

**Comment by sobolevnrm** [comment]
Hello --

How large (dimensions in Angstroms) is your system?  

Running with the `async` keyword sounds like your easiest option:  you just run a sequence of `mg-para` calculations with `async` set from 0 to 299.  You should not need an open-mpi build of APBS for this option.

Given the variation in MPI flavors, MPI-enabled versions of APBS need to be built from source.

Thank you,

Nathan

**Comment by srdjan125** [comment]
Hello,

Thank you for answer.

The dimensions of system and parameters recommended by PDB2PQR website (or psize.py program) are:

   read
    mol pqr AB_15nm.pqr
end
elec
    mg-auto
    dime 481 609 321
    cglen 377.0787 487.5328 238.2788
    fglen 241.8110 306.7840 160.1640
    cgcent mol 1
    fgcent mol 1


or for parallel calculations(also from PDB2PQR website):

read
mol pqr 3umojvbirp.pqr
end
elec
mg-para
dime 481 609 321
pdime 6.0 6.0 3.0
ofrac 0.1
cglen 377.0787 487.53279999999995 238.27880000000002
fglen 241.811 306.784 160.16400000000002
cgcent mol 1
fgcent mol 1
mol 1

 May I get some help how can I run several mg-para calculations with async option? I could not understand it from tutorial, but maybe it exists in some another tutorial or forum? If I could see some example it would easier for me to understand.
There is inputgen.py program which should create several mg-para+async files but I am having difficulty to run this program.

  As I understand, mg-para with async option means running one calculation, and output of that calculations serves like input to next mg-para(async) calculation?
 But how many "series" of these mg-para(async) calculations I should perform? Does cglen and fglen in those calculations remain the same?   In my case recommended number of processors is:  pdime 6.0 6.0 3.0 . Does it mean that I need to set up async rank 108(=6*6*3)      Or I should put maximum number of cores which I can use on my cluster (async rank 32) ?

  If I would be able to see some tutorial (similar to apbs/examples) it would be much easier for me to prepare those calculations.

   Thank you for help, greetings.


________________________________

**Reply by sobolevnrm** [reply]
Hello.  The `pdime` keyword requires 3 integers.  In your case, that leads to 108 processors, as you pointed out.  If you wanted to run the trivially parallel job serially, you would set up a series of calculations, each with increasing `async` rank; e.g.:  `async 0`, `async 1`, ..., `async 107`.  These jobs will not require MPI -- you can just submit them to your cluster's queue and let the scheduler sort them out.

More information about this command can be found at https://apbs.readthedocs.io/en/latest/using/input/elec/async.html?highlight=async#async

**Comment by srdjan125** [comment]
Hello,

Thank you for the answer. In the meantime I have figured out async calculations: I just wrote python program to write input files wuth async 0. async 1 etc, then run them sequentally and at the end use python to sum values from all individual async calculations (as async performs calculation at one part of box).
  I just forgot to update it into this task.

   May I ask one final question: Let's say I need 108 processors for parallel run, but I can have only 5-6 processors maybe (or maximum 10). Does it mean that I could invoke mpi run with only 10 processors, even in mg-para asks for 108 processors? It would still be possible to run it just the speed would not be optimal?

   Or I should just stick with async option?

   Thank you for help, greetings.



________________________________

**Reply by sobolevnrm** [reply]
I would just stick with the `async` option.

Thank you,

Nathan
